#160577579 Enable Article Like and Dislike

imireallan opened the pull request:
## What does this PR do?
The PR enables authenticated user to appreciate articles by either liking or disliking them.

## Description of Task to be completed?

##### Why this is important
Lets appreciate those authors who go out of their way to create good content with a Like and encourage those authors who do not meet our expectations with a dislike.

##### As a User
I want to be able to like or dislike articles  articles, both authored by me and those not authored by me.

##### Acceptance Criteria
Scenario:
Only **authenticated** users should be able to like or dislike articles.


## How can it be manually tested?
Clone the repository to your local environment.
```
git clone https://github.com/andela/ah-codeblooded-frontend.git
```

Change directories
```
cd ah-codeblooded-frontend/
```

Install the dependencies
```
yarn install
```

Run the server
```
yarn start
```

### It will open a browser with the current route at `/` and view the list of articles.

### Backend setup

- Authenticate a user on using the `ah-code-blooded` backend 
- Add the user credentials in the browser local storage using this this format:
```
{
        "username" : "<username>",
        "bio" : "<bio>",
        "avatar" : "",
        "token" : <token>
}
```
### Test Like/Dislike
- View one of the pre-existing articles on the home page
- Scroll to the bottom of the article and you will see the "like" and "dislike" button icons as shown on the screenshot below

**Screenshots (if appropriate)**
![image](https://user-images.githubusercontent.com/31212441/48483786-7211c000-e825-11e8-9591-decd68afc0ee.png)


**What are the relevant pivotal tracker stories?**

The relevant pivotal story is [here](https://www.pivotaltracker.com/n/projects/2198535/stories/160577579)

Imire committed:
feat(like/dislike): implement environment variables

- ensure all sensitive information is secured in the environment

[Starts #160577579]

Imire committed:
feat(like/dislike): implement axios api module

- ensure axios calls are modularized

[Starts #160577579¬]

Imire committed:
feat(article-like-dislike): implement article like and dislike

- created like-dislike reducer
- created actions for like-dislike

[Starts #160577579]

SrMoffat commented on src/containers/LikeDislike/index.js:
```diff
@@ -0,0 +1,31 @@
+// import React, { Component } from 'react';
```
Hey @imireallan,
How come this entire file has been commented out?

imireallan replied:
@SrMoffat this is because the pr as at now only shows the functionality of reducers and actions the commented file is for the presentational component which am currently doing

SrMoffat commented on src/utils/api.js:
```diff
@@ -0,0 +1,52 @@
+
+
+// const token = getToken();
+const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJpbWlyZSIsImVtYWlsIjoiaW1pcmVAZ21haWwuY29tIiwiaWF0IjoxNTQxNTgyMjQzLCJleHAiOjE1NDE1ODk0NDN9.KqR2wO0Q94VjC3zr7CqwYYz-_6HrQY0j63jiBUILlcw';
```
Hey @imireallan ,
I'm sure this `token` is just a mockup of the server behaviour right?

imireallan replied:
yes @SrMoffat am using that way because the login functionality is yet to be finished

paulrimiru replied:
This Token should be in the environment variables to avoid it being pushed to the repo

imireallan replied:
@PaulKariukiRimiru the token is just for testing because the sign up and login is yet to be merged

SrMoffat reviewed (commented):
Hey @imireallan,
Nice work so far. I would like to point out that you need to update your PR description under `Manual Testing` because it currently has the generic text from the template. Additionally, I have also included other observations in the file comments. Other than that, if you fix the `Hound` issues and and the `CRUD` for articles is done, I believe we can be able to test the functionality and see.

Imire committed:
feat(article-like-dislike): add tests for likeDislikeReducer

- tests the reducer has an initial state
- tests the reducer can dispatch actions

[Starts #160577579'

charisschomba reviewed (commented):
Good work @imireallan , consider include api tests.

Imire committed:
feat(article-like-dislike): adds a reaction component

- implements reaction component
- ensures the reaction component will be implemented in the read article

[Starts #160577579]

paulrimiru reviewed (changes requested):
The setup section of your description is not complete, review it

paulrimiru commented on .env_example:
```diff
@@ -0,0 +1 @@
+REACT_APP_API_URL=https://ah-code-blooded-staging.herokuapp.com
```
The URL should not be pushed to the repo.

imireallan replied:
working on it..

paulrimiru commented on src/components/Reactions/index.js:
```diff
@@ -0,0 +1,24 @@
+import PropTypes from 'prop-types';
+
+const Reactions = (props) => {
+  const { likes, dislikes, handleClick } = props;
```
Make sure you add prop validations

paulrimiru commented on src/components/Reactions/index.js:
```diff
@@ -0,0 +1,24 @@
+  );
+};
+
+// eslint-disable-next-line react/no-typos
```
Add this rule to the eslint config file instead. It makes sense but not as an inline rule exception but as part of the global rules

paulrimiru commented on src/containers/LikeDislike/state/Actions.js:
```diff
@@ -0,0 +1,66 @@
+
+import api from '../../../utils/api';
+
+
```
Remove extra line here

imireallan replied:
done

paulrimiru commented on src/containers/LikeDislike/state/Actions.js:
```diff
@@ -0,0 +1,66 @@
+import api from '../../../utils/api';
+
+
+export const likeArticle = slug => (dispatch) => {
```
Please make the param more descriptive `slug` is not descriptive enough

imireallan replied:
how should this be? is articleSlug ok?

paulrimiru commented on src/containers/LikeDislike/state/likeDislikeReducer.js:
```diff
@@ -0,0 +1,36 @@
+import {
```
I dont completely understand your thought process on structuring the store like this, would you please explain your thought process?

imireallan replied:
i have updated the structure you can review and comment

paulrimiru commented on src/containers/LikeDislike/state/likeDislikeReducer.test.js:
```diff
@@ -0,0 +1,87 @@
+  LIKE_ARTICLE, DISLIKE_ARTICLE, UNDISLIKE_ARTICLE, UNLIKE_ARTICLE,
+} from './types';
+
+describe('like/dislike reducer', () => {
```
Test cases description should read as a sentence, For example, the description of a test suite should be The beginning of the sense eg `The Like and Dislike reducer` then each test suite should add the remaining part of the sentence.

imireallan replied:
updated

paulrimiru commented on src/containers/LikeDislike/state/likeDislikeReducer.test.js:
```diff
@@ -0,0 +1,87 @@
+  });
+  it('should handle LIKE_ARTICLE', () => {
+    expect(likeDislikeReducer(undefined, {
+      type: LIKE_ARTICLE,
```
Declare this object as a global variable and reuse it throughout, to make the code more DRY

imireallan replied:
hey, do u mean the action object?

imireallan replied:
worked on it

paulrimiru commented on src/utils/api.js:
```diff
@@ -0,0 +1,52 @@
+  Authorization: `token ${token}`,
+};
+
+const api = {
```
Make this axios instance more reusable, use an axions instance for this for example
```
const axiosInstance = axios.create({
  baseURL: {{url}},
  headers: {
    Authorization: `Bearer ${token}`,
  },
});
```

imireallan replied:
updated this

paulrimiru reviewed (changes requested):
Fix the codeclimate issues and improve the test coverage

Imire committed:
feat(article-like-dislike): implement like-dislike container

- implement a container for the like/dislike
- refactor tests to ensure they pass with the new changes

[Starts #160577579]

Imire committed:
feat(article-like-dislike): add tests for the redux like/dislike container

- adds more tests for like/dislike actions
- adds tests for the redux like/dislike container

[Starts #160577579]

paulrimiru commented on src/containers/LikeDislike/likeDislike.test.js:
```diff
@@ -0,0 +1,93 @@
+    // expect(spy.calledOnce).toBe(true);
+  });
+
+  // it('should map state to props', () => {
```
Remove this test if it is not required

imireallan replied:
cool..working on it

paulrimiru commented on src/containers/LikeDislike/state/action.test.js:
```diff
@@ -0,0 +1,145 @@
+  const dislikeAction = payload => ({ type: DISLIKE_ARTICLE, payload });
+  const undislikeAction = payload => ({ type: UNDISLIKE_ARTICLE, payload });
+
+  // const likeAction = {
```
REmove this code if it is not necessary

paulrimiru commented on src/containers/LikeDislike/state/action.test.js:
```diff
@@ -0,0 +1,145 @@
+    });
+  });
+
+  it('should dispatch DISLIKE_ARTICLE', () => {
```
Separate test cases with a new line

imireallan replied:
ok

paulrimiru reviewed (changes requested):
Good work, add tests for failure actions and also fix the issues I highlighted then we can merge this.

charisschomba commented on src/containers/LikeDislike/state/Actions.js:
```diff
@@ -0,0 +1,66 @@
+import {
+  LIKE_ARTICLE, LIKE_DISLIKE_ERROR, DISLIKE_ARTICLE, UNLIKE_ARTICLE, UNDISLIKE_ARTICLE,
+} from './types';
```
Consider importing each of these in a new line.

charisschomba commented on src/containers/LikeDislike/state/action.test.js:
```diff
@@ -0,0 +1,178 @@
+import api, { getURL } from '../../../utils/api';
+import {
+  FETCH_REACTIONS_SUCCESS, LIKE_ARTICLE, DISLIKE_ARTICLE,
+  UNDISLIKE_ARTICLE, UNLIKE_ARTICLE, LIKE_DISLIKE_ERROR,
```
same case here.

charisschomba commented:
Please solve conflicting files above

charisschomba commented:
@imireallan It seems that this PR was not deployed.
<img width="579" alt="screen shot 2018-11-14 at 23 31 43" src="https://user-images.githubusercontent.com/40663486/48510672-9e4d3100-e865-11e8-825c-73d78fc00886.png">

imireallan commented:
> Please solve conflicting files above

on it

mutaimwiti reviewed (approved):
Looks good to me.
